packages/@react-aria/button/src/useButton.ts

```diff
@@ -91,3 +94,18 @@ export function useButton(props: AriaButtonProps, ref: RefObject<HTMLElement>):
     })
   };
 }
+
+export function useToggleButton(props: AriaToggleButtonProps, state: ToggleState, ref: RefObject<HTMLElement>): ButtonAria {
+  const {isSelected, setSelected} = state;
+  const {isPressed, buttonProps} = useButton({
+    ...props,
+    onPress: chain(() => setSelected(!isSelected), props.onPress)
```

Comment:
In general, if we're going to use the state while we're setting the new value, we should make sure not to use the wrong value which can happen since state updates are batched. Instead, if we pass a function to set'State' we're guaranteed to have the latest value
```suggestion
    onPress: chain(() => setSelected(prev => !prev), props.onPress)
```

Reply:
on a separate thought, it's strange that useToggleState doesn't return a toggle function... @devongovett how do you feel about adding that, we can do it in a different PR, this just made me think of it

Reply:
Yeah we should add that

Reply:
@snowystinger setSelected() does not take a function so the best I could do is 
```typescript
onPress: chain(() => setSelected(!state.isSelected), props.onPress)
```

Reply:
if I get the 2nd comment right you suggest the following change:
```typescript
export interface ToggleState {
  /** Whether the toggle is selected. */
  readonly isSelected: boolean,

  /** Updates selection state. */
  setSelected(isSelected: boolean): void
  
  /** Toggle the selection state */
  toggle(): void
}

/**
 * Provides state management for toggle components like checkboxes and switches.
 */
export function useToggleState(props: ToggleProps = {}): ToggleState {
  let {isReadOnly, onChange} = props;

  // have to provide an empty function so useControlledState doesn't throw a fit
  // can't use useControlledState's prop calling because we need the event object from the change
  let [isSelected, setSelected] = useControlledState(props.isSelected, props.defaultSelected || false, () => {});

  function updateSelected(value) {
    if (!isReadOnly) {
      setSelected(value);
      if (onChange) {
        onChange(value);
      }
    }
  }

  return {
    isSelected,
    setSelected: updateSelected,
    toggle: () => { setSelected(!isSelected) }
  };
}
```

Reply:
pretty close, but that won't cover fast updates, readonly, or notify onChange
if you change the toggle line like this, that should cover it. (though you'll probably want to move it out to a similar function like `updateSelected`)
```
toggle: () => {
  if (!isReadOnly) {
      setSelected(prev => {
        let newVal = !prev;
        if (onChange) {
          onChange(newVal);
        }
        return newVal;
      });
  }
}
```

Reply:
here, you might find this helpful https://reactjs.org/docs/hooks-reference.html#functional-updates

Reply:
Cool - I think I've done it now the right way. I guess the only thing missing now are the test cases.